Enable pushing Sort/Filter by ReferenceAttribute down to Lucene, and thereby optimize Sort by ST_DISTANCE

craigtaverner opened the pull request:
The `ST_DISTANCE` function added in https://github.com/elastic/elasticsearch/pull/108764 was optimized for lucene pushdown in a series of followup PRs, but this did not include sorting by distance. Now this is resolved, for two key scenarios, both known to be valued by users:

* Sorting by distance:
  * `FROM index | EVAL distance=ST_DISTANCE(field, literal) | SORT distance`
* Sorting and filtering by distance:
  * `FROM index | EVAL distance=ST_DISTANCE(field, literal) | WHERE distance < literal | SORT distance`

The key changes required to make this work:

* Add to the `EsQueryExec` the appropriate `sort->_geo_distance` sort type
* Enhance `PushTopNToSource` to understand how to pushdown the sort even when there is an `EVAL` in between the `FROM` and the `SORT` (between the `TopNExec` and the `EsQueryExec` in the physical plan).
* Enhance `PushFiltersToSource` to understand how to pushdown the filter even when there is an `EVAL` in between the `FROM` and the `WHERE` (between the `FilterExec` and the `EsQueryExec` in the physical plan).

A useful bonus feature of this additional `EVAL` intelligence is that other, non-spatial cases are now also pushed down. In particular `EVAL`s that are simple aliases (`ReferenceAttribute`s to `FieldAttribute`s) are considered and pushed down, for both filtering and sorting.

Local benchmark results, very approximate, but show massive improvements for `distanceSort` and `distanceFilterSort`, which relate to the two cases listed above.

| Benchmark | Query DSL | ESQL before this PR | ESQL after this PR | Comments |
| --- | --- | --- | --- | --- |
| distanceFilter | 10 | 5 | 5 |  Optimized in https://github.com/elastic/elasticsearch/issues/109972  |
| distanceEvalFilter | 10 | 10000 | 1500 | [Still slow due to unnecessary EVAL](https://github.com/elastic/elasticsearch/issues/114310) |
| distanceSort | 150 | 12000 | 160 |    |
| distanceFilterSort | 20 | 10000 | 24 |    |

*NOTE:* This enables pushing down sorting by any `ReferenceAttribute` that either refers to a sortable `FieldAttribute`, or to an `StDistance` function that itself refers to a suitable `FieldAttribute` of `geo_point` type.

Fixes #109973

Checklist:

- [x] Code prototype of happy path (FROM->EVAL(dist)->SORT(dist) combination) with integration tests verifying behaviour (behaviour driven development to drive walking skeleton)
- [x] Support TransportVersion change using trivial name string to separate FieldSort from GeoDistanceSort
- [x] PhysicalPlanOptimizerTests to understand scope of the requirements and how far the prototype goes:
  - [x] Tests for sort pushdown without spatial (eg. `FROM index | SORT age`) - *PASSES*
  - [x] Tests for sort pushdown without spatial but with filters (eg. `FROM index | WHERE age < 20 | SORT age`) - *PASSES*
  - [x] Tests for sort pushdown with spatial (eg. `FROM index | EVAL distance=ST_DISTANCE | SORT distance`) - *PASSES*
  - [x] Tests for sort pushdown with spatial and filter (eg. `FROM index | EVAL distance=ST_DISTANCE | WHERE distance < 50000 | SORT distance`) - *PASSES*
  - [x] Tests for sorting on both plain fields and evaluated fields (eg. distance) in same command (eg. FROM index | EVAL distance=ST_DISTANCE | SORT age, distance`) - *PASSES*
  - [x] Tests for filtering and sorting on both plain fields and evaluated fields (eg. distance) in same command (eg. FROM index | EVAL distance=ST_DISTANCE | WHERE age > 20 AND distance < 50000 | SORT age, distance`) - *PASSES*
- [x] Support pushdown of spatial filter using EVAL(distance), since existing pushdown requires function inside the WHERE clause
- [x] Support combined SORT with both fields and EVAL results (distance or trivial aliases)
- [x] More unit tests for EsQueryExec.Sort (see existing tests for EsQueryExec.FieldSort)
- [x] ~Consider improving TransportVersion change to use NamedWritable instead of hand coded string to separate FieldSort from GeoDistanceSort~ (removed GeoDistanceSort serialization, to be followed up in https://github.com/elastic/elasticsearch/issues/113809)
- [x] Write unit tests for PushTopNToSource, since it is now much more complex than before
- [x] Support multiple evals: esp `FROM index | EVAL point=... | EVAL distance=ST_DISTANCE(loc, point) | SORT distance`
- [x] Generalize SORT pushdown to allow EVAL in more cases than spatial distance
  - [x] Trivial aliasing: `FROM index | EVAL x = y | SORT x`
  - [ ] ~Linear functions: `FROM index | EVAL x = y + 1 | SORT x`~ (too complex to consider at this point, unlikely to be considered later)
- [ ] Followup work:
  - [ ] Simplify the rules as much as possible, with a more general solution for aliasing - Can all trivial aliasing be moved out of the current rules
  - [ ] https://github.com/elastic/elasticsearch/issues/114310
  - [ ] Validate this works with multi-values - https://github.com/elastic/elasticsearch/issues/112910
  - [ ] https://github.com/elastic/elasticsearch/issues/113809
  - [x] https://github.com/elastic/elasticsearch/issues/114577
  - [x] https://github.com/elastic/elasticsearch/issues/114515
  - [ ] Consider generalizing to other optimizable patterns, like [this one](https://github.com/elastic/elasticsearch/pull/112938#discussion_r1785948613).
  - [ ] Push down in case of [aliases inside distances](https://github.com/elastic/elasticsearch/pull/112938#discussion_r1792206614).
  - [ ] Address final minor remarks [here](https://github.com/elastic/elasticsearch/pull/112938#discussion_r1793233641) and from [this review](https://github.com/elastic/elasticsearch/pull/112938#pullrequestreview-2362358065).

craigtaverner committed:
ES|QL Sort by ST_DISTANCE pushdown to Lucene

This prototype works on single-instance servers, because plan serialization has not yet been fixed to include the new type of sorting in the EsQueryExec serialization.

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Added transportversion for more generic EsQueryExec sort type

craigtaverner committed:
Support SORT and WHERE pushdown with EVAL distance

Since SORT requires an EVAL, we needed to also support WHERE pushdown on EVAL in order to combine SORT and WHERE pushdown in the same query.

This has the side effect of allowing a few more non-spatial cases to support SORT pushdown if they are preceded by an EVAL that is irrelevant to the SORT. We could expand this to support EVALS that do impact the sort in a trivial way (eg. sort on an alias).

craigtaverner committed:
Expand doc-values support to StDistance

Was only supported on SpatialRelatesFunction, which did not cover StDistance, but if StDistance is used together with ST_CENTROID_AGG, we get a failure.

craigtaverner committed:
Pushing down more TopN to lucene changes profile output

craigtaverner committed:
More unit tests for EsQueryExec.GeoDistanceSort

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/EsQueryExec.java:
```diff
@@ -56,8 +59,19 @@ public class EsQueryExec extends LeafExec implements EstimatesRowSize {
-    public record FieldSort(FieldAttribute field, Order.OrderDirection direction, Order.NullsPosition nulls) implements Writeable {
-        public FieldSortBuilder fieldSortBuilder() {
+    public interface Sort {
+        SortBuilder<?> sortBuilder();
```
I think we should not be using DSL objects here. Could not we use here a lucene SortField instead?

craigtaverner replied:
As I understand it we'll consider that later, so I'll mark this as resolved.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,168 @@
+            // a distance function defined in the EVAL. We also move the EVAL to after the SORT.
+            List<Order> orders = topNExec.order();
+            List<Alias> fields = evalExec.fields();
+            LinkedHashMap<NameId, StDistance> refs = fields.stream().reduce(new LinkedHashMap<>(), (m, a) -> {
```
I am wondering if you considered `AttributeMap` usage here instead of this.

craigtaverner replied:
I'm looking into that, but it does not seem like a correct match for my needs. I need to lookup from ReferenceAttributes found later to the StDistance function within the Alias that the ReferenceAttribute points to. However there is no attribute available at map creation, only the Alias.id() which is later referred to. I could imagine doing a two-pass algorithm over the orders and the aliases to try build an AttributeMap, but that would require this intermediate map anyway. The current approach seems to be a much simpler solution.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java:
```diff
@@ -161,15 +161,15 @@ public Function<org.elasticsearch.compute.lucene.ShardContext, Query> querySuppl
-        List<FieldSort> sorts = esQueryExec.sorts();
-        List<SortBuilder<?>> fieldSorts = null;
+        List<Sort> sorts = esQueryExec.sorts();
+        List<SortBuilder<?>> sortBuilders = null;
```
I would have kept EsQueryExec the way it was and deal with geo-specific sorting here instead. This class is dealing with the translation of physical plan objects to actual operators (elements that do the actual physical work).
The LocalExecutionPlanner initiates the planning and coordinates the different physical plan components transformation and the PhysicalOperationProviders implementation deal with the creation of the operators for each specific physical plan. Having the specific implementation leak into EsQueryExec feels wrong.

craigtaverner replied:
It feels like it already leaked into EsQueryExec. The fact that it contained a `FieldSort` which was specific to sorting on raw fields only, means that it was already not general to other kinds of sorts. If we did not add the geo sort info to EsQueryExec, we would need to attach it to something somewhere, and I cannot see where else it could be attached.

craigtaverner replied:
I was thinking that perhaps what you are looking for is a way to not change the serialization of EsQueryExec, by keeping the GeoDistanceSort evaluation lower down, as you said _"here"_ in the `EsPhysicalOperationProviders`. The problem is that we can only push this down if we have the combination `TopNExec->EvalExec->EsQueryExec` with all the right conditions. This tree structure is not possible to determine here, and the only code that looks at the plan tree, and figures these things out is the `LocalPhysicalPlanOptimizer`, which is why we put it there.

craigtaverner committed:
Wrote unit tests for PushTopNToSource

These tests are designed to test the various combinations of TopNExec(EvalExec(EsQueryExec))) where pushing to source will work, and where it will not work. This combination of positive and negative tests is designed to complement the integration tests which focus on positive full-stack tests.

craigtaverner committed:
Test inverted distance and foldable distance functions

craigtaverner committed:
Less strict on error message for heap-attack

This PR causes the memory error to occur with different operators, usually one of:

* `[<esql_block_factory>]`
* `[serialize exchange response]`

Instead of the previous `[topn]`.

craigtaverner committed:
Remove all traces of random testing in unit test

This is because the random values have zero impact on the tests, and their presence gives the false impression that they matter. Rather use deliberately invalid values like `null` where possible so it is more obvious that this content is irrelevant to these unit tests.

craigtaverner committed:
More thorough check on specific planned classes

craigtaverner committed:
Trivial cleanup on review

astefan reviewed (commented):
I've taken a deeper look at the PR and I have some comments:
- the use cases it covers are very targeted and minimal; is it really worth the effort (which is not trivial) to allow pushdown in a very limited set of cases?! I know there there is the expectation of a followup, but tbh I would have expected a more general approach to be covered. For example, `FROM airports | eval x = 1 | EVAL distance = ST_DISTANCE(location, TO_GEOPOINT(\"POINT(12.565 55.673)\")) | WHERE distance <= 600000 AND distance >= 400000` is not covered (the simple presence of `x=1` there bypasses the optimization)
- 👍 for use cases where we didn't push down regular fields before. The title of the PR doesn't reflect that, though
- `FROM index | WHERE age < 20 | SORT age` is this always correct (regarding the pushdown)? Combined with the implicit limit we always add

craigtaverner commented:
Some answers:

> * the use cases it covers are very targeted and minimal; is it really worth the effort (which is not trivial) to allow pushdown in a very limited set of cases?! I know there there is the expectation of a followup, but tbh I would have expected a more general approach to be covered. For example, `FROM airports | eval x = 1 | EVAL distance = ST_DISTANCE(location, TO_GEOPOINT(\"POINT(12.565 55.673)\")) | WHERE distance <= 600000 AND distance >= 400000` is not covered (the simple presence of `x=1` there bypasses the optimization)

Spatial distance filtering and sort combined is the most used spatial feature in Elasticsearch, so that combination was the primary target. I did notice today that some trivial changes to this by splitting to multiple EVAL's did not get the optimization (eg. `FROM airports | EVAL point=TO_GEOPOINT("POINT(12.565 55.673)") | EVAL distance=ST_DISTANCE(location, point) | WHERE  distance <= 600000 AND distance >= 400000 | SORT distance | LIMIT 10`), and agree that this should be done, and might be a tiny change to this PR, or a bigger change for a followup PR. This would cover the case of the `EVAL x=1` you suggest above too.

> * 👍 for use cases where we didn't push down regular fields before. The title of the PR doesn't reflect that, though

I could update the title, but that enhancement seemed like a tiny improvement compared with the main work on distance pushdown.

> * `FROM index | WHERE age < 20 | SORT age` is this always correct (regarding the pushdown)? Combined with the implicit limit we always add

I believe this already worked before this PR. The filter was pushed down first, and the WHERE clause removed, and then the SORT got pushed down on the next pass of the local physical planning. I added more tests for this because I wanted to verify the behaviour, but this was pre-existing behaviour. I did have a problem with WHERE clauses and ST_DISTANCE, which I did fix in this PR; but for normal WHERE clauses like the one you describe here, the pushdown of both filter and sort was already working.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java:
```diff
@@ -80,6 +83,42 @@ public void writeTo(StreamOutput out) throws IOException {
+     * This is because the planner might push down a spatial aggregation to lucene, which results in the field being provided
+     * as doc-values instead of source values, and this function needs to know if it should use doc-values or not.
+     */
+    public boolean hasFieldAttribute(Set<FieldAttribute> foundAttributes) {
```
These methods pre-exist this PR, and were just moved from one class to another to fix an issue with the scope of the spatial filter pushdown. However, I see your point that they could be moved to the `SpatialDocValuesExtraction` class since that is the only place they are ever used. This could be done in the PR; or in another cleanup PR.

astefan replied:
> These methods pre-exist this PR
Yes, I've tracked down the history of those methods, but now I took my time to look over the code concerning spatial functions in detail.

Let's try doing the move in this PR 👍

craigtaverner replied:
Done!

astefan commented:
> and agree that this should be done, and might be a tiny change to this PR, or a bigger change for a followup PR

The change could be small, but not easy I think. To be done right, it needs a bit of thinking on the variety of use cases covered. Follow up PR, I'd say.

> I added more tests for this because I wanted to verify the behaviour, but this was pre-existing behaviour.

Thank you for adding the tests. I really appreciate extending them for such cases of optimizations. The more, the better, as we cover edge cases.

Regarding [the aspect of `EsQueryExec` leakage](https://github.com/elastic/elasticsearch/pull/112938#discussion_r1773661915), after re-reviewing yesterday, I believe the change is ok 👍

craigtaverner committed:
Fix spatial distance filter pushdown on multiple evals

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner committed:
Allow sort pushdown for aliased fields

craigtaverner committed:
Moved some physical planning specific code out of BinarySpatialFunction

craigtaverner committed:
Added unit tests for distance pushdown with aliased fields

craigtaverner committed:
The previous fix for field aliases was trivial to extend for field sorts

craigtaverner committed:
Support pushing filters to source with reference attributes to field attributes

And in particular support pushing distance to source if both the distance and the literal points are defined in EVALS.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+                }
+            });
+            List<EsQueryExec.Sort> pushableSorts = new ArrayList<>();
+            for (Order order : orders) {
+                if (LucenePushDownUtils.isPushableFieldAttribute(order.child(), hasIdenticalDelegate)) {
+                    pushableSorts.add(
+                        new EsQueryExec.FieldSort(
+                            ((FieldAttribute) order.child()).exactAttribute(),
+                            order.direction(),
+                            order.nullsPosition()
+                        )
+                    );
```
See my previous comment, pick on the sort then move backwrds - instead of expecting an eval then a sort.
that might be common scenario however there might be minor variations that could mke this to fail.

craigtaverner replied:
I'm not sure I follow. This class seems to do what you suggest. Start with the sort (TopNExec) and then look backwards (to child->child, etc.). I do differentiate between a simple TopNExec->EsQueryExec, which is sufficient for a sort-by-field query, and a more advanced TopNExec->EvalExec->EsQueryExec, which is *required* for the distance sort because we cannot sort on expressions, only attributes. This PR does, however, allow sorting on trivial aliases too, like `FROM index | EVAL x = y | SORT x | LIMIT 10`.

craigtaverner replied:
As far as I can see, this PR only expands the range of queries that can be pushed down. Each time you describe a case that might `fail`, it seems more like a case that might not get the optimization. I do not (yet) see an actual failure.

costin reviewed (commented):
Left some comments around the rules - let me know if you'd like to chat more about them.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java:
```diff
@@ -80,6 +81,29 @@ public void writeTo(StreamOutput out) throws IOException {
+    /**
+     * Mark the function as expecting the specified fields to arrive as doc-values.
+     */
+    public abstract BinarySpatialFunction withDocValues(boolean foundLeft, boolean foundRight);
```
This code pre-exists this PR by about half a year. It just got moved up to the parent class, because there was a bug with ST_DISTANCE and centroid aggregations. Having said that, I also cannot see how to move this out of the plan nodes, because this is knowledge generated during planning that is then used to influence field extraction later. The knowledge of whether we use doc values is determined by the planning rule `SpatialDocValuesExtraction` which needs to look at a tree of physical plan nodes to determine if the right combination exists to take this decision, and then mark several nodes in the plan with this knowledge, notably the centroid aggregation itself, and any spatial functions that might be affected by this (any functions called before the centroid aggregation).

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
-        if (filterExec.child() instanceof EsQueryExec) {
+        if (filterExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, esQueryExec);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, evalExec, esQueryExec);
+        }
```
The "looking for sort commands" part is done by a different rule, see the `PushTopNToSource` rule. However, for that rule to work, filters (WHERE clauses) that can also be pushed down need to be pushed down first. This logic pre-dates my PR, and was the reason why pushing down sorts worked together with filters. I wrote a bunch of additional physical plan optimizer tests to verify this behaviour, and those tests worked before any of my changes. So my work here is just expanding this to ST_DISTANCE, which needs to be pushed down in the WHERE clause as well as the SORT (two different pushdowns, but both need to exist). I had a basic WHERE pushdown in the original ST_DISTANCE PR for 8.15, but it turns out I missed a few important use cases. So the current PR enhances the WHERE pushdown in `EnableSpatialDistancePushdown` rule as well as adds new SORT pushdown in the `PushTopNToSource` rule. It keeps the same thinking as before, which is that the sort pushdown does not need to know or care about the WHERE pushdown, and leaves it to this rule.

Regarding multiple EVALs, this is handled in the code further down, and my physical plan optimization tests cover WHERE clauses preceded by multiple EVALs. Also keep in mind that this class does not actually perform pushdown, as hinted by the class name, it only `enables` and existing pushdown by re-writing the `ST_DISTANCE` into `ST_INTERSECTS` and `ST_DISJOINT` which are pushdown down later. This is also the way that the existing Elasticsearch REST API does distance search.

craigtaverner replied:
I've investigated and indeed, while my code for pushing sort down does deal with deeply nested aliases, my code here for pushing filters to source does not. So I've pushed a commit that fixes that. I also did look into using AttributeMap in the sort-pushdown code, but it was not appropriate there. I'll re-evaluate its usage for this case and see if it makes more sense here. There are now three places where I do reference attribute lookups, so I'll re-evaluate all three.

craigtaverner committed:
Allow nested aliases for distance filter pushdown

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner committed:
Remove sort from serialization of EsQueryExec, without TransportVersions change

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -43,55 +46,108 @@
-                    exp
-                );
+            plan = planFilterExec(filterExec, queryExec, ctx);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec) {
+            plan = planFilterExec(filterExec, evalExec, queryExec, ctx);
+        }
```
```
FROM index
EVAL poi = TO_GEOPOINT("POINT (1 2)")
EVAL distance = ST_DISTANCE(location, poi)
WHERE distance > 10000 AND distance < 50000
```

I have a bunch of similar and more complex examples in the PhysicaPlanOptimizerTests as well as in benchmarks.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+            // When we have an exchange between the FROM and the SORT, we also only allow pushing down if the sort is on a field
+            return new PushableExchangeExec(exchangeExec, queryExec);
+        }
+        if (child instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec && queryExec.canPushSorts()) {
+            // When we have an EVAL between the FROM and the SORT, we consider pushing down if the sort is on a field and/or
+            // a distance function defined in the EVAL. We also move the EVAL to after the SORT.
+            List<Order> orders = topNExec.order();
+            List<Alias> fields = evalExec.fields();
+            LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
+            LinkedHashMap<NameId, FieldAttribute> refAttributes = new LinkedHashMap<>();
+            fields.forEach(alias -> {
+                // TODO: can we support CARTESIAN also?
+                if (alias.child() instanceof StDistance distance && distance.crsType() == BinarySpatialFunction.SpatialCrsType.GEO) {
+                    refs.put(alias.id(), distance);
+                } else if (alias.child() instanceof FieldAttribute fieldAttribute) {
+                    refAttributes.put(alias.id(), fieldAttribute);
+                } else if (alias.child() instanceof ReferenceAttribute ref && refAttributes.containsKey(ref.id())) {
+                    FieldAttribute fieldAttribute = refAttributes.get(ref.id());
+                    refAttributes.put(alias.id(), fieldAttribute);
+                }
+            });
+            List<EsQueryExec.Sort> pushableSorts = new ArrayList<>();
```
Yes, after your last comment, there are now three places where we trace ReferenceAttributes. I will re-review all three and see if:
* AttributeMap can be used
* The three places can share code

In the previous place, the map was a one-way tree down to StDistance, simple, but in this case we are dealing with two leaf types, FieldAttribute and StDistance (referring to FieldAttribute). And in both cases, the reference can be indirect. So sharing code and/or using AttributeMap seem unlikely. But I'll double-check.

craigtaverner committed:
Extensive java-doc on spatial functions use of doc-values.

Questions have been asked a few times about why the spatial functions contain state that is not serialized, and is only specific to knowledge on local data nodes. This java-doc hopefully answers that.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner committed:
Added a simpler version of an existing test to help investigate

Remaining performance concerns in the benchmarks.
Benchmarks where the filter is within the WHERE clause are much faster than those where it is in a preceding EVAL. It looks like the performance issue is that the EVAL remains even after the distance filter is pushed down, even if it is no longer required.

craigtaverner committed:
Removed println from new test

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+    }
+
+    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
+        LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
```
Normally we use AttributeMap for this (it's also just a map based on name id equality), but that's also fine.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+        return refs;
+    }
+
+    private PhysicalPlan rewrite(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
```
I was thinking about situations like
```
FilterExec
\_EvalExec
  \_FilterExec
    \_EsSourceExec
```
where both filters could be pushable. We do not account for this here, but would that be a relevant case?

We could get that from
```
FROM idx | WHERE ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) <= 10000 | EVAL dist = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) | WHERE dist > 300
```
or so.

If cases like that are relevant, Costin's comment will become more relevant to expand the situations where the push down works. I.e. rather than looking for a certain pattern, we'll have to look for FilterExecs where substituting reference attributes from previous evals leads to a pushable condition.

But let's keep it focused for now :)

craigtaverner replied:
I would have thought that consolidating EVAL's and WHERE's into fewer, combined commands (in another rule) could resolve that, but can also imagine that might cause performance issues in some cases (eg. the first WHERE filters down the data so the later EVAL has less work). In any case, I agree that we should investigate and consider cases like this in a later PR. This PR is supposed to tackle performance optimization of a number of key use cases, and now already optimizes more cases than initially envisaged, so I'm happy to push more complex cases like this to later work.

alex-spies replied:
I just tried and my example from above gets optimized away in 2 rounds of `PushFiltersToSource`. But the situation from above can be forced with a non-pushable filter, like in
```
FROM idx | WHERE ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) <= 10000 AND field::string == "asdf" | EVAL dist = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) | WHERE dist > 3000
```

craigtaverner replied:
I think this scenario could be handled in a followup PR.

alex-spies reviewed (commented):
I had a first look, will take another one tomorrow!

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
     @Override
     protected PhysicalPlan rule(FilterExec filterExec, LocalPhysicalOptimizerContext ctx) {
```
Good point. I've added new content to the javadoc to describe the new scenario where we get the distance from an EVAL instead of directly within the WHERE clause.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java:
```diff
@@ -80,6 +81,29 @@ public void writeTo(StreamOutput out) throws IOException {
+    public int hashCode() {
+        // NB: the hashcode is currently used for key generation so
+        // to avoid clashes between aggs with the same arguments, add the class name as variation
+        return Objects.hash(getClass(), children(), leftDocValues, rightDocValues);
```
This is covered by the getClass method, since pointOnly is true for one class and false for all others. While we could express pointsOnly differently (eg. a method on the class, so more obviously tied to the class itself), I would rather do that in a different PR since it is unrelated to the purpose of this PR (ie. a pre-existing concern). Since this was a trivial refactoring that did not change the behaviour of equals/hashcode, just moved it up to the parent class, lets consider that later. My preference would be when we implement distance functions for shapes, which will require a change to this logic anyway.

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -1116,9 +1116,42 @@ count:long | country:k
 required_capability: st_distance
 
+FROM airports
+| EVAL distance = ST_DISTANCE(location, TO_GEOPOINT("POINT(12.565 55.673)"))
+| SORT distance ASC
```
I thought we could only sort on attributes, not expressions. I'll investigate.

craigtaverner replied:
The docs claim we can only sort on columns (attributes).

craigtaverner replied:
It seems to work, so I've added the test and fixed a bug that blocked this behaviour for spatial types.

craigtaverner committed:
Added comments reflecting increased scope

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner committed:
Support sorting on distance function directly

craigtaverner committed:
Verify that sort over inline distance function is pushed down

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -552,10 +552,9 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
-                if (DataType.isSpatial(dataType)) {
-                    localFailures.add(fail(attr, "cannot sort on " + dataType.typeName()));
+            ob.order().forEach(order -> {
+                if (DataType.isSpatial(order.dataType())) {
+                    localFailures.add(fail(order, "cannot sort on " + order.dataType().typeName()));
```
++!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +68,70 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
-                plan = new FilterExec(filterExec.source(), filterExec.child(), rewritten);
+                // Rewrite the order so the EVAL is last, allowing the filter to be pushed down to lucene by PushFiltersToSource
+                // TODO: Remove the distance attribute from the EVAL if it is no longer needed
+                FilterExec filter = new FilterExec(filterExec.source(), esQueryExec, rewritten);
+                return new EvalExec(evalExec.source(), filter, evalExec.fields());
```
This looks like a bug and I got a reproducer:
```
FROM test | EVAL dist = ST_DISTANCE(x, TO_GEOPOINT(\"POINT(0 0)\")), foo = x::string | WHERE dist > 3000 AND foo == \"asdf\"

{"error":{"root_cause":[{"type":"null_pointer_exception","reason":"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"}],"type":"null_pointer_exception","reason":"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null","suppressed":[{"type":"exception","reason":"1 further exceptions were dropped"},{"type":"task_cancelled_exception","reason":"parent task was cancelled [cancelled on failure]"}]},"status":500}
```
It's not always possible to move the whole filter down. We must carefully remove only the pushable predicates from the filter.

Looking at the effects of the optimization rule shows the problem quite nicely:
```
[2024-10-03T11:56:03,927][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.EnableSpatialDistancePushdown applied
ProjectExec[[x{f}#40, dist{r}#34, foo{r}#37]]                                                                                  = ProjectExec[[x{f}#40, dist{r}#34, foo{r}#37]]
\_LimitExec[1000[INTEGER]]                                                                                                     = \_LimitExec[1000[INTEGER]]
  \_FilterExec[dist{r}#34 > 3000[INTEGER] AND foo{r}#37 == [61 73 64 66][KEYWORD]]                                             !   \_EvalExec[[STDISTANCE(x{f}#40,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist, TOSTRING(x{f}#40) AS foo]]
    \_EvalExec[[STDISTANCE(x{f}#40,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist, TOSTRING(x{f}#40) AS foo]] !     \_FilterExec[SPATIALDISJOINT(x{f}#40,[1 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 70 a7 40][GEO_SHAPE]) AND foo{r}
      \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#41], limit[], sort[] estimatedRowSize[null]                    ! #37 == [61 73 64 66][KEYWORD]]
                                                                                                                               !       \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#41], limit[], sort[] estimatedRowSize[null]

```
The results of the EVAL might still be needed for other predicates.

craigtaverner replied:
I've fixed this bug in https://github.com/elastic/elasticsearch/pull/112938/commits/27a8b8cf62dad23fbd825a0bed81a7e88f1317c3

alex-spies reviewed (changes requested):
Found a :bug:  where the optimization breaks a query that used to work before (I think).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -43,55 +46,108 @@
         }
+        // Replace field references with their actual field attributes
+        pushable.replaceAll(e -> e.transformDown(ReferenceAttribute.class, r -> fieldAttribute(r, refs)));
+        return rewrite(filterExec, queryExec, pushable, nonPushable);
```
I don't think this can be right.

This disregards the EvalExec altogether. But the EvalExec might evaluate values that are used downstream, or it might evaluate non-pushable predicates for the FilterExec.

craigtaverner replied:
I believe this is resolved in https://github.com/elastic/elasticsearch/pull/112938/commits/27a8b8cf62dad23fbd825a0bed81a7e88f1317c3. I maintain also the other aliases that need to be maintained in the EvalExec.

alex-spies reviewed (commented):
Found another thing that I think needs addressing, this time in PushFiltersToSource. The problem is similar to the previous one: we assume that in the `Source -> Eval -> Filter` chain, the Eval only computes values for pushable filters, and disregard that it could compute values for non-pushable filters, or for other consumers down stream.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/EsQueryExec.java:
```diff
@@ -56,8 +59,19 @@ public class EsQueryExec extends LeafExec implements EstimatesRowSize {
+
+        FieldAttribute field();
+
+        void writeTo(StreamOutput out) throws IOException;
```
I wonder if we really need to serialize this? EsQueryExec as a whole might never be serialized.

craigtaverner replied:
I wanted to leave the complete removal of serialization support from EsQueryExec to a separate PR; and so I kept serialization in but made it all dummy serialization. However, your point is valid that some of the remaining code is no longer required since I made the serialization of the Sort all dummy serialization (ie. empty lists). I still need to de-serialize, because we can be receiving from an old node, but at least the `writeTo` methods can go.

craigtaverner replied:
Done.

alex-spies reviewed (commented):
Ok, I reviewed all non-test code now. No new major remarks, so to get this in I think we should focus on the bugs in the `Query -> Eval -> Filter` cases.

craigtaverner committed:
Fix bug with spatial filter pushdown when eval has other filtered attributes

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +68,70 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+                // TODO: Remove the distance attribute from the EVAL if it is no longer needed
+                FilterExec filter = new FilterExec(filterExec.source(), esQueryExec, rewritten);
+                return new EvalExec(evalExec.source(), filter, evalExec.fields());
             }
```
I fixed this in https://github.com/elastic/elasticsearch/pull/112938/commits/27a8b8cf62dad23fbd825a0bed81a7e88f1317c3, where the EVAL is split into two depending on which parts are needed in the Filter. I also agree that we should consider moving this level of logic to the logical planner. There are consequences to doing that, so I want to only attempt that in a followup PP.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -43,55 +46,108 @@
-                );
+            plan = planFilterExec(filterExec, queryExec, ctx);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec) {
+            plan = planFilterExec(filterExec, evalExec, queryExec, ctx);
```
This deals with deeply nested aliases, while the other class focuses on StDistance only. The combination was required to resolve all the use cases I was after. Having this logic here also adds some very nice features for non-spatial cases, like `FROM index | EVAL x = y | WHERE x > 10`. But at the end of the day the split I was after was handle aliasing in this class, and handle StDistance->Intersects mapping in the other class. When we investigate moving logic to the logical planner, I suspect this split will guide us (ie. I think the StDistance->Intersects mapping more naturally moves up to logical planning).

craigtaverner committed:
Code review updates, notably use of AttributeMap instead of custom maps in two places.

craigtaverner committed:
Comment grammar

craigtaverner committed:
Added a comment and removed an unused method

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -168,29 +224,38 @@ else if ((other instanceof GreaterThan || other instanceof GreaterThanOrEqual)
+
+    public static boolean canPushToSource(
+        Expression exp,
+        Map<NameId, FieldAttribute> refs,
```
Refactored to use AttributeMap.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -201,6 +266,40 @@ public static boolean canPushToSource(Expression exp, Predicate<FieldAttribute>
+            return isPushableSpatialAttribute(refs.get(ref.id()));
+        }
+        return isPushableSpatialAttribute(exp);
+    }
```
Refactored to use AttributeMap.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+            if (attr instanceof FieldAttribute fieldAttribute) {
+                return fieldAttribute;
+            } else if (attr instanceof ReferenceAttribute ref && refAttributes.containsKey(ref.id())) {
+                return refAttributes.get(ref.id());
```
Refactored to use AttributeMap.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            var sorts = buildFieldSorts(topNExec.order());
+            var limit = topNExec.limit();
+            return exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
```
Good question. This part of the code pre-existed my work, and so I kept in in place, but have to be honest I think it is surprising. Perhaps I just delete it, and see if no tests fail, and assume test coverage is sufficient? But since it pre-existed, and does not relate to my work, I decided to just leave it in place.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+            // When we have an exchange between the FROM and the SORT, we also only allow pushing down if the sort is on a field
+            return new PushableExchangeExec(exchangeExec, queryExec);
+        }
+        if (child instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec && queryExec.canPushSorts()) {
```
The `EnabledSpatialDistancePushdown` does a re-write from `ST_DISTANCE` to `ST_INTERSECTS`, and then relies on `PushFiltersToSource` to do the real pushdown. However, for `PushTopNToSource` there is no special distance->intersects rewriting going on, the distance sort is pushed down as is, so we did not consider moving that out to a separate rule, and kept it here.

In addition, we are now considering moving some logic to the logical planner, and `EnabledSpatialDistancePushdown` is a clear candidate, but most of what happens in either `PushFiltersToSource` or `PushTopNToSource` is much less clear, and more likely to remain in local physical planning.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -157,6 +166,11 @@ public ShapeRelation relation() {
+
+        @Override
+        public String toString() {
+            return Strings.toString(this, true, true);
```
Yes. I did add something similar in some other work a few months back, but removed it to keep that PR clean. This time I wanted to keep it, because those exceptions were just too irritating.

craigtaverner committed:
Update x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/SpatialDocValuesExtraction.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

craigtaverner committed:
Removed unused Writable methods from EsQueryExec.Sort classes

craigtaverner committed:
Merge branch 'st_distance_sort_pushdown' of github.com:craigtaverner/elasticsearch into st_distance_sort_pushdown

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,204 @@
+
+        private static PushableGeoDistance from(
+            Attribute attr,
+            Expression literal,
```
I could do that, but then would need to split the `makeGeometryFromLiteral` into two versions, one for pushdown and one for compute engine. Since Literal is foldable() and fold() just calls value(), this way seems cleaner and simpler. Also, I did not want to make this code sensitive to exactly when literal folding occurs, so we could move it around later, more easily.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -94,46 +95,82 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
-                StDistance distance = refs.get(ref.id());
-                refs.put(alias.id(), distance);
+            if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance)) {
+                distances.put(alias.id(), distance);
```
The standard way is to have an `AttributeMap` and then `distances.put(alias.toAttribute(), distance)` rather than using the name id directly.

Let's refactor this - if needed in a direct follow-up - since I think it's important to have just 1 way to do things in the optimizer for simplicity's sake.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -94,46 +95,82 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+                if (comparison.left() instanceof ReferenceAttribute r && distances.containsKey(r.id()) && comparison.right().foldable()) {
+                    StDistance dist = distances.get(r.id());
                     return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
                 } else if (comparison.right() instanceof ReferenceAttribute r
```
The fact that we rewrite only in case of reference attributes looks incorrect. I think in an execution plan like
```
Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(renamed_field{r}, POINT(1 -1)) > 5]
\_Eval[renamed_field = other_field{f}]
  \_Query
```
this will rewrite only the predicate on the `renamed_field` but leave the non-renamed predicate on `field` alone.

I also do not believe that the best approach to deal with aliases that only rename is to keep them before the filter and splitting the eval. I believe we should split the filter and keep the eval alone.

In more detail: let's update the filter by replacing reference attributes by field attributes - then part of the filter can be brought closer to the source. (In fact, that's something that should happen in the logical planning, but let's keep it locally here for this PR and move out in a follow-up.)

The reason why I think that's important is that bringing filters close to the source is something we do in general, and it allows us to focus on combinations `Query -> Filter` for Lucene pushdowns. I'd like to have as little special casing as possible for cases like `Query -> Eval -> Filter`, because looking for patterns consisting of 3 plan nodes is less general and more brittle than looking for patterns consisting of 2 plan nodes.

For example, I think the plan from above should be rewritten to
```
Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(other_field{f}, POINT(1 -1)) > 5]
\_Eval[renamed_field = other_field{f}]
  \_Query
```
and then the filter can even be wholly pushed down
```
Eval[renamed_field = other_field{f}]
\_Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(other_field{f}, POINT(1 -1)) > 5]
  \_Query
```
So that we only need to replace the `st_distance`s and then `PushFiltersToSource` can just look at the `Query -> Filter` pattern at the bottom and push the filter further down.

In contrast, the rule currently doesn't seem to rewrite this plan at all because the eval is needed for the filter - and thus `PushFiltersToSource` needs to look for the `Query -> Eval -> Filter` pattern.

alex-spies replied:
Never mind. Filters that don't have reference attributes that refer to expressions defined in the Eval should have been pushed down past the Eval in the first place, so the situation I described doesn't need to be handled here.

alex-spies reviewed (commented):
Found another bug that needs addressing before we can merge.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,212 @@
+            return null;
+        }
+
+        private static PushableGeoDistance from(Attribute attr, Expression literal, Order order) {
```
nit:
```suggestion
        private static PushableGeoDistance from(Attribute attr, Expression foldable, Order order) {
```
otherwise this suggests that only `Literal` instances should be used, but in fact foldability is all we need.

Similarly, I think `makeGeometryFromLiteral` should be renamed to `makeGeometryFromFoldable`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,212 @@
+                        )
+                    );
+                } else if (order.child() instanceof ReferenceAttribute referenceAttribute) {
+                    if (distances.containsKey(referenceAttribute.id())) {
```
nit: This doesn't account for multiple aliasing, as in `EVAL dist = st_distance(x, "POINT(0 0)"::geo_point), dist2 = dist, dist3 = dist2 | SORT dist3 DESC`.

I think the correct way to do this is to 1. resolve the ref attribute using `aliasReplacedBy` (AttributeMap should take multiple indirections into account), and then 2. check if what we resolved to is either a pushable field attr or one of the pushable distances.

This would also do away with the code duplication in lines 168 and 189.

craigtaverner replied:
Did a two line fix to cover this case (and expanded and existing test to test it). Did not reduce code duplication, because we cannot resolve the alias without first checking the type, and having that type check is kind-of the code duplication part. Perhaps some cleanup is possible, but not a complete de-dupliciation.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java:
```diff
@@ -7,55 +7,212 @@
+                        } else {
+                            // If the SORT refers to a non-pushable reference function, the EVAL must remain before the SORT,
+                            // and we can no longer push down anything
+                            break;
+                        }
+                } else {
+                    // As soon as we see a non-pushable sort, we know we need a final SORT command
+                    break;
```
Bug! I think in these two cases, we should `return NO_OP` instead of `break`ing - in case we find one pushable sort but then the next sort is not pushable.

Reproducer:
```
FROM test | sort st_distance(x, \"POINT(0 0)\"::geo_point) asc, x::keyword desc
```
This pushes down the distance sort but ignores the sort on `x::keyword`:
```
[2024-10-09T11:36:07,239][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.PushTopNToSource applied
ProjectExec[[x{f}#20, $$order_by$0$1{r}#21, $$order_by$1$2{r}#22]]                                                          = ProjectExec[[x{f}#20, $$order_by$0$1{r}#21, $$order_by$1$2{r}#22]]
\_TopNExec[[Order[$$order_by$0$1{r}#21,ASC,LAST], Order[$$order_by$1$2{r}#22,DESC,FIRST]],1000[INTEGER],null]               ! \_EvalExec[[STDISTANCE(x{f}#20,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS $$order_by$0$1, TOSTRING(x{f}#2
  \_EvalExec[[STDISTANCE(x{f}#20,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS $$order_by$0$1, TOSTRING(x{f}#2 ! 0) AS $$order_by$1$2]]
0) AS $$order_by$1$2]]                                                                                                      !   \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#23], limit[1000], sort[[GeoDistanceSort[field=x{f}#20, direction=ASC, lat=0.0, lon=0.0]]] estimatedRowSize[null]
    \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#23], limit[], sort[] estimatedRowSize[null]                   ! 

```

alex-spies reviewed (commented):
Got another one.

craigtaverner committed:
Fixed bug where attempt to rewrite distance for pushdown leaves references in filter

This could be better fixed by splitting the filter, but that will be attempted in followup since it is a more comprehensive rewrite.

craigtaverner committed:
Code review - don't mutate 'others' as code is hard to follow

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +69,106 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+            List<Alias> referencedAliases = new ArrayList<>();
+            filterExec.condition().forEachDown(ReferenceAttribute.class, r -> {
+                if (others.containsKey(r.id())) {
+                    referencedAliases.add(others.remove(r.id()));
```
OK. Pushed a commit. The code is now a little more complex, but does not mutate others.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +69,106 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+            // If there are remaining aliases, we need to keep them after the filter
+            List<Alias> remainingAliases = new ArrayList<>(others.values());
+            // Add back the distance functions, since they might be used in later clauses
+            // TODO: Remove this if we can guarantee that the distance functions are only used in the filter
```
I agree this is not trivial or local, but thought the TODO was useful as a reminder to at least think about this as part of possible followup work.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +69,106 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+            // Determine if the filter refers to any of the other fields in the EVAL
+            List<Alias> referencedAliases = new ArrayList<>();
+            filterExec.condition().forEachDown(ReferenceAttribute.class, r -> {
+                if (others.containsKey(r.id())) {
```
Fixed by double-checking that the filter rewrite actually removes all references to the distance function, not merely that it changes something (which was insufficient a condition).

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -53,45 +59,94 @@
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            (canPushToSource(resExp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(
+                exp
```
I'll wait for the other simplification coming down the line rather than making this change from ternary to `if...else` since we probably don't need the reference attribute mapping at all once that is in place.

craigtaverner committed:
Added alternative approach to distance pushdown, by splitting filters

This approach does not split the EVAL as before, but splits the filter, moving the pushable part down, and the non-pushable part after the eval.

This commit does not remove the older approach. That will be done in a separate commit.

craigtaverner committed:
Cleaned up and removed code for splitting eval instead of filter

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -4882,14 +4882,16 @@ public void testPushTopNDistanceAndNonPushableEvalWithCompoundFilterToSource() {
-        var alias = as(evalDistance.fields().get(0), Alias.class);
-        assertThat(alias.name(), is("distance"));
-        var stDistance = as(alias.child(), StDistance.class);
+        var filter = as(topNChild.child(), FilterExec.class);
```
We should probably assert the filter's condition.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -95,80 +99,105 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+            // Find and rewrite any binary comparisons that involve a distance function and a literal
+            var rewritten = rewriteDistanceFilters(resExp, distances);
+            // If all pushable StDistance functions were found and re-written, we need to re-write the FILTER/EVAL combination
+            if (rewritten.equals(resExp) == false && conditionHasNoReferences(rewritten, distances)) {
```
Isn't `conditionHasNoReferences(rewritten, distances)` always true and thus redundant if the rewrite was succesful?

craigtaverner replied:
That fix originally fixed the conjunctive `WHERE distance < 50000 AND 0*distance < 40000`, but while the new split-filter should also fix that, I think we need this for disjunctive cases, like `WHERE distance < 50000 OR 0*distance < 40000`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -95,80 +99,105 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
                 distances.put(alias.id(), distance);
             } else if (alias.child() instanceof ReferenceAttribute ref && distances.containsKey(ref.id())) {
                 StDistance distance = distances.get(ref.id());
                 distances.put(alias.id(), distance);
```
Since we rely on `AttributeMap.resolve()` now and that takes renames into account, I think this case is now  obsolete.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -95,80 +99,105 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+     *     | WHERE other &gt; 10
+     * </pre>
+     */
+    private PhysicalPlan rewriteBySplittingFilter(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
```
Nice, looks correct now!

alex-spies reviewed (commented):
We're heading in a good direction!

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -53,45 +59,94 @@
+                queryExec.estimatedRowSize()
+            );
+            // If the eval contains other aliases, not just field attributes, we need to keep them in the plan
+            PhysicalPlan plan = others.isEmpty() ? queryExec : new EvalExec(filterExec.source(), queryExec, others);
```
Indeed. Once I found a failing test, I fixed this, and then found some existing tests that failed because they did not actually assert on the right thing, so this was a bug affecting more cases than I realized. Our integration tests did not notice because they usually have a KEEP, and that always modifies the plan at logical planning time to avoid this bug. But queries with EVAL and no KEEP quite easily hit this.

craigtaverner committed:
Fixed bug with dropping eval fields during filter pushdown

This was only affecting queries that did not have an explicit KEEP in them, because that KEEP caused a rewrite of the filter aliases anyway. So queries affected needed the FROM->EVAL->WHERE combination where the WHERE references attributes defined in the EVAL, and there is no later KEEP command (regardless of whether it mentions those fields or not).

craigtaverner committed:
Tests with disjunctive WHERE

This test will fail if we remove the `conditionHasNoReferences` check from `EnableSpatialDistancePushdown`

craigtaverner commented on docs/changelog/112938.yaml:
```diff
@@ -0,0 +1,35 @@
+
+    [source,esql]
+    ----
+    FROM test
+    | EVAL ref = field
+    | SORT ref ASC
+    | LIMIT 10
```
I understand that on main this will not be pushed to Lucene, but run in the compute engine. The query works, but is now optimized.

alex-spies replied:
Ok, the wording makes it sound like this wasn't supported before, so maybe let's update the wording :)

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
I see I have a test for deeply nested distance and SORT, but not for WHERE, so I could cover this case with WHERE later.

craigtaverner committed:
Update x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java:
```diff
@@ -44,23 +74,125 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
-                plan = new FilterExec(filterExec.source(), filterExec.child(), rewritten);
+            var rewritten = rewriteDistanceFilters(resExp, distances);
+            // If all pushable StDistance functions were found and re-written, we need to re-write the FILTER/EVAL combination
+            if (rewritten.equals(resExp) == false && conditionHasNoReferences(rewritten, distances)) {
```
Now that I thought a bit more about it, I think this condition is too coarse. In particular, we are pushing down the distance computation even if we cannot push down to Lucene in some cases:
```
FROM test | eval dist = st_distance(x, \"POINT(0 0)\"::geo_point) | where dist > 5 OR x::keyword == \"asdf\"

[2024-10-10T15:02:59,037][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.EnableSpatialDistancePushdown applied
ProjectExec[[x{f}#14, y{f}#15, dist{r}#11]]                                                                 = ProjectExec[[x{f}#14, y{f}#15, dist{r}#11]]
\_LimitExec[1000[INTEGER]]                                                                                  = \_LimitExec[1000[INTEGER]]
  \_FilterExec[dist{r}#11 > 5[INTEGER] OR TOSTRING(x{f}#14) == [61 73 64 66][KEYWORD]]                      !   \_EvalExec[[STDISTANCE(x{f}#14,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist]]
    \_EvalExec[[STDISTANCE(x{f}#14,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist]]        !     \_FilterExec[SPATIALDISJOINT(x{f}#14,[1 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 14 40][GEO_SHAPE]) OR TOSTRING
      \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#16], limit[], sort[] estimatedRowSize[null] ! (x{f}#14) == [61 73 64 66][KEYWORD]]
                                                                                                            !       \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#16], limit[], sort[] estimatedRowSize[null]
```
There was no subsequent run of `PushFiltersToSource` that would've done anything about the filter - so we made the query worse by computing the distance twice!

craigtaverner replied:
Since the issue was that we re-wrote even if the next PushFiltersToSource would fail to push the rewritten filter, the general condition here was to use `canPushToSource` on the rewritten filter, which returns false if it contains any component that blocks the pushdown. So this is now fixed, and there is a text verifying it.

The old `conditionHasNotReferences` is deleted, as it was too specific, as you said.

craigtaverner committed:
Added TODO for removing ExchangeExec support from PushTopNToSource

craigtaverner committed:
Support sorting on aliases distances

craigtaverner committed:
Merge branch 'st_distance_sort_pushdown' of github.com:craigtaverner/elasticsearch into st_distance_sort_pushdown

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -53,47 +59,96 @@
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            (canPushToSource(resExp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(
```
Very subtle bug: in case we don't push the attribute, we mustn't add it to the non pushables!

Since `resExp` comes with resolved reference attributes, it may point to attributes that existed before the eval but are shadowed in the eval. Like so:
```
test has fields x and y:

FROM test | EVAL y = 2*y, dist = st_distance(x, "POINT(0 0)") | WHERE dist < 10 AND y > 1
```
For `y > 1`, resExp uses a different `y` which does not exist anymore after the Eval!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -53,47 +59,96 @@
+        EsQueryExec queryExec,
+        List<Expression> pushable,
+        List<Expression> nonPushable,
+        List<Alias> others
```
nit:
```suggestion
        List<Alias> evalFields
```

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -53,47 +59,96 @@
+                exp
+            );
+        }
+        // Replace field references with their actual field attributes
+        pushable.replaceAll(e -> e.transformDown(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r)));
```
Isn't the resolution duplicated here? We already do that in line 96.

craigtaverner replied:
No, we added the non-resolved one to the collection, and so need to resolve it again. This can be changed by expanding the ternary to an `if...else`, but I've been trying to leave that ternary alone...

The similar code in the EnableSpatialDistancePushdown does it with `if..else` and only one resolution. I considered merging the two bits of code, but decided that did not improve clarity.

craigtaverner committed:
Fix compile error from partial commit

craigtaverner committed:
Fixed last two bugs, partial sorts, and compound unpushable distance filters

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

craigtaverner committed:
Update x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

craigtaverner committed:
Compile error after merging a commit

craigtaverner committed:
Added test to verify that non-pushable filters do not affect pushable filters

craigtaverner committed:
Shadow the name to verify this has no impact

craigtaverner committed:
Remove println

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlannerTests.java:
```diff
@@ -107,6 +107,21 @@ public void testLuceneTopNSourceOperator() throws IOException {
+    public void testLuceneTopNSourceOperatorDistanceSort() throws IOException {
+        int estimatedRowSize = randomEstimatedRowSize(estimatedRowSizeIsHuge);
+        FieldAttribute sortField = new FieldAttribute(Source.EMPTY, "point", new EsField("point", DataType.GEO_POINT, Map.of(), true));
+        EsQueryExec.GeoDistanceSort sort = new EsQueryExec.GeoDistanceSort(sortField, Order.OrderDirection.ASC, 1, -1);
```
nit: we could randomize the sort a little for good measure (at least the sort direction between ASC and DESC)

alex-spies reviewed (approved):
Nice! I think the changes to the optimization rules are very good now and this can be merged. I may leave a couple comments after this review, but if so that'll be something for a follow-up.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_distance_sort_pushdown

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/114604)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/114604">](https://github.com/elastic/elasticsearch/pull/114604) |

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is not labelled `release highlight` but the changelog has a `highlight` section

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -3826,6 +4214,1361 @@ AND ST_DISTANCE(location, TO_GEOPOINT("POINT(12.565 55.673)")) >= 200000))
+        var topNChild = as(extract.child(), TopNExec.class);
+        extract = as(topNChild.child(), FieldExtractExec.class);
+        assertThat(names(extract.attributesToExtract()), contains("scalerank"));
+        var filter = as(extract.child(), FilterExec.class);
```
nit: we could assert that the filter only has the predicate `SUBSTRING(position, 1, 5) == "POINT"` remaining, as the rest is pushed down.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -3826,6 +4214,1361 @@ AND ST_DISTANCE(location, TO_GEOPOINT("POINT(12.565 55.673)")) >= 200000))
+    public void testPushCompoundTopNDistanceWithDeeplyNestedCompoundEvalToSource() {
+        var optimized = optimizedPlan(physicalPlan("""
+            FROM airports
+            | EVAL poi = TO_GEOPOINT("POINT(12.565 55.673)")
+            | EVAL poi2 = poi, poi3 = poi2
+            | EVAL loc2 = location
+            | EVAL loc3 = loc2
+            | EVAL dist = ST_DISTANCE(loc3, poi3)
+            | EVAL distance = dist
+            | SORT scalerank, distance
+            | LIMIT 15
```
I think we should have a similar test for filter pushdown. But I think it currently will fail due to EnableSpatialDistancePushdown not resolving refs inside distances.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -3826,6 +4214,1361 @@ AND ST_DISTANCE(location, TO_GEOPOINT("POINT(12.565 55.673)")) >= 200000))
+        var optimized = optimizedPlan(physicalPlan("""
+            FROM airports
+            | EVAL poi = TO_GEOPOINT("POINT(12.565 55.673)")
+            | EVAL distance = ST_DISTANCE(location, poi)
```
nit: tests could randomize `ST_DISTANCE(location, poi)` vs `ST_DISTANCE(poi, location)` (and similarly in the other tests), as we have the distinction between the 2 cases in multiple places. Or we eliminate the problem by canonicalizing the expression in our optimization rules.

alex-spies reviewed (commented):
I have a couple final, minor remarks on the tests that we could address in the follow-ups :)

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is not labelled `release highlight` but the changelog has a `highlight` section
